Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm

```diff
@@ -227,27 +260,40 @@
         return;
     }
 
-    auto replacedRangeAfterReplace = WebCore::resolveCharacterRange(*sessionRange, characterRangeAfterReplace, defaultTextAnimationControllerTextIteratorBehaviors);
+    auto previouslyReplacedRange = m_alreadyReplacedRanges.getOptional(sessionUUID);
+
+    auto replacedCharacterRange = characterRangeAfterReplace;
+    if (previouslyReplacedRange)
+        replacedCharacterRange = remainingCharacterRange(characterRangeAfterReplace, *previouslyReplacedRange);
+    auto replacedRangeAfterReplace = WebCore::resolveCharacterRange(*sessionRange, replacedCharacterRange, defaultTextAnimationControllerTextIteratorBehaviors);
 
     auto unstyledRange = WebCore::makeRangeSelectingNodeContents(*document);
     unstyledRange.start.container = replacedRangeAfterReplace.endContainer();
     unstyledRange.start.offset = replacedRangeAfterReplace.endOffset();
 
     auto unstyledRangeUUID = WTF::UUID::createVersion4();
     TextAnimationUnstyledRangeData unstyledRangeData = { unstyledRangeUUID, unstyledRange };
-    m_unstyledRanges.add(sessionUUID, unstyledRangeData);
+    m_unstyledRanges.set(sessionUUID, unstyledRangeData);
 
     auto textIndicatorData = createTextIndicatorForRange(replacedRangeAfterReplace);
     if (!textIndicatorData)
         return;
 
-    m_webPage->addTextAnimationForAnimationID(finalTextIndicatorUUID, { TextAnimationType::Final, unstyledRangeUUID }, *textIndicatorData);
-    TextAnimationState animationState = { finalTextIndicatorUUID, characterRangeAfterReplace };
-    auto& animationStates = m_activeTextAnimations.ensure(sessionUUID, [&] {
-        return Vector<TextAnimationState> { };
+    m_webPage->addTextAnimationForAnimationID(destinationTextIndicatorUUID, { TextAnimationType::Final, unstyledRangeUUID }, *textIndicatorData, [weakWebPage = WeakPtr { *m_webPage }, sessionUUID]() mutable {
+        if (!weakWebPage)
+            return;
+        WTFLogAlways("<MMG> calling the completion Handler!!! you did it!");
+        weakWebPage->corePage()->chrome().client().addInitialTextAnimation(sessionUUID);
```

Comment:
It's a bit weird for WebKit (client) code to need to call into the `Chrome`/`ChromeClient` just to propagate back up to the client layer.

Maybe simpler if you just call into `WebPage` directly?

```suggestion
        weakWebPage->addInitialTextAnimation(sessionUUID);
```